stl/inc/mutex

```diff
@@ -35,19 +35,22 @@ class condition_variable_any;
 
 class _Mutex_base { // base class for all mutex types
 public:
-#ifdef __CUDACC__ // TRANSITION, CUDA
-    _Mutex_base(int _Flags = 0) noexcept : _Mtx_storage(_Flags | _Mtx_try) {
-        _Mtx_init_in_situ(_Mymtx(), _Flags | _Mtx_try);
-    }
-#else // ^^^ TRANSITION, CUDA ^^^ / vvv not CUDA vvv
-    constexpr _Mutex_base(int _Flags = 0) noexcept : _Mtx_storage(_Flags | _Mtx_try) {
-        if (!__builtin_is_constant_evaluated()) {
-            // If mutex is created in runtime, it may outlive the module it is allocated in;
-            // Use usual in-place construction.
+#if _HAS_CXX20
+    constexpr
+#endif // _HAS_CXX20
+        _Mutex_base(int _Flags = 0) noexcept {
+#if _HAS_CXX20
+        if (_STD is_constant_evaluated()) {
+            ::new (static_cast<void*>(&_Mtx_storage)) _Stl_critical_section_constexpr{};

```

Comment:
This _new-expression_ isn't valid in a constant expression since it doesn't select a replaceable global allocation function ([expr.const]/5.19). We _could_ use `construct_at`, but it will need storage for an actual `_Stl_critical_section_constexpr` to target - presumably in a union, which (I think) will make `mutex` not standard-layout.

Reply:
> This _new-expression_ isn't valid in a constant expression since

What if I use custom form of placement new, that would be constexpr?

> which (I think) will make `mutex` not standard-layout

Right, we can't do union here.

Reply:
Addressed, seem to compile

Reply:
> > This _new-expression_ isn't valid in a constant expression since
> 
> What if I use custom form of placement new, that would be constexpr?

Either it wouldn't really allocate storage ("select a replaceable global allocation function") or it wouldn't deallocate the storage within the evaluation of the constant expression (also [expr.const]/5.19).

Reply:
Got your point. But what if the non-standard code still compiles?

Reply:
Ok, I put `construct_at` there, but with a `static_cast`. Is this still a crime?

Reply:
That violates [expr.const]/5.15 since it involves "a conversion from type *cv* `void*` to a pointer-to-object type".